sgxs-loaders/src/enclaveapi/mod.rs

```diff
@@ -0,0 +1,268 @@
+        device: Arc<WinInnerLibrary>,
+        ecreate: MeasECreate,
+        attributes: Attributes,
+        miscselect: Miscselect,
+    ) -> Result<Mapping<Self>, Self::Error> {
+        let secs = Secs {
+            size: ecreate.size,
+            ssaframesize: ecreate.ssaframesize,
+            miscselect,
+            attributes,
+            ..Default::default()
+        };
+
+        let curhandle: HANDLE = unsafe {GetCurrentProcess()};
+        let base = unsafe {
+            CreateEnclave(
+                curhandle,
+                ptr::null_mut(),
+                ecreate.size as _,
+                0,
+                ENCLAVE_TYPE_SGX,
+                &secs as *const _ as *const c_void,
+                mem::size_of::<Secs>() as _,
+                ptr::null_mut()
+            )
+        };
+
+        if base.is_null() {
+            Err(Error::Create(IoError::last_os_error()))
+        } else {
+            Ok(Mapping {
+                device,
+                tcss: vec![],
+                base: base as _,
+                size: ecreate.size,
+            })
+        }
+    }
+    fn add(
+        mapping: &mut Mapping<Self>,
+        page: (MeasEAdd, PageChunks, [u8; 4096]),
+    ) -> Result<(), Self::Error> {
+
+        let (eadd, chunks, data) = page;
+
+        if eadd
+            .secinfo
+            .flags
+            .intersects(SecinfoFlags::PENDING | SecinfoFlags::MODIFIED | SecinfoFlags::PR)
+        {
+            return Err(Error::Add(ErrorKind::InvalidInput.into()));
+        }
+        let mut flags : winapi::shared::minwindef::DWORD = 0;
+        match (eadd.secinfo.flags & (SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X)).bits() as u8 {
+            0b000 => {},
+            0b001 => flags = PAGE_READONLY,
+            0b010 => {return Err(Error::Add(ErrorKind::InvalidInput.into()))},
+            0b011 => flags = PAGE_READWRITE,
+            0b100 => flags = PAGE_EXECUTE,
+            0b101 => flags = PAGE_EXECUTE_READ,
+            0b110 => {return Err(Error::Add(ErrorKind::InvalidInput.into()))},
+            0b111 => flags = PAGE_EXECUTE_READWRITE,
+            0b1000...255 => unreachable!(),
+        }
+
+        match PageType::from_repr(eadd.secinfo.flags.page_type()) {
+            Some(PageType::Reg) => {}
+            Some(PageType::Tcs) => {
+                if eadd.secinfo.flags.contains(SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X) {
+                    return Err(Error::Add(ErrorKind::InvalidInput.into()))
+                }
+                // NOTE: For some reason the windows API needs the Read flag set but then removes it
+                flags = PAGE_ENCLAVE_THREAD_CONTROL|PAGE_READWRITE;
+            },
+            _ => return Err(Error::Add(ErrorKind::InvalidInput.into())),
+        }
+        match chunks.0 {
+            0 => flags = flags | PAGE_ENCLAVE_UNVALIDATED,
+            0xffff => {}
+            _ => return Err(Error::Add(ErrorKind::InvalidInput.into())),
```

Comment:
`IoError::new(ErrorKind::Other, "Partially-measured pages not supported on Windows")`